### PR TITLE
Compat mode. 

### LINKED ISSUE #775: react-stand-in incorrectly picks up Relay containers as functional components
This is against v4 beta 13.

Relay Classic uses an odd pattern to lazily build containers: https://github.com/facebook/relay/blob/d77a34c6b5abddfe078b0d50c2d43181fe35d574/packages/react-relay/classic/container/RelayContainer.js#L1050-L1055

This behaves as a normal class, but does not get picked up as a React class here.

### LINKED ISSUE #832: `Objects are not valid as a React child` after upgrading `react-stand-in` from `4.0.0-beta.13`
Moved from https://github.com/theKashey/react-stand-in/issues/2

### Description

What you are reporting: Bug

### Expected behavior

What you think should happen: do not throw error

### Actual behavior

What actually happens:

```
Error: Objects are not valid as a React child (found: object with keys {props, context, refs, updater, __reactstandin__proxyGeneration, __reactstandin__cachedResult}). If you meant to render a collection of children, use an array instead.null
    at invariant (invariant.js:42)
    at traverseAllChildrenImpl (react.development.js:805)
    at traverseAllChildren (react.development.js:833)
    at Object.countChildren [as count] (react.development.js:945)
    at doTraverse (index.js:84)
    at doVisit (index.js:116)
    at index.js:172
    at new Promise (<anonymous>)
    at reactTreeWalker (index.js:70)
    at mapper (index.js:88)
```

### Environment

React Hot Loader version: 4.0.0-beta.18
react-stand-in version: >= 4.0.0-beta.14

Run these commands in the project folder and fill in their results:

1. `node -v`: v9.4.0
2. `npm -v`: 5.6.0

Then, specify:

1. Operating system: macOS
2. Browser and version: Chrome 64

### Reproducible Demo

https://github.com/JounQin/react-ssr/tree/react-stand-in

Branch: react-stand-in

For now, I'm locking react-stand-in@4.0.0-beta.13, you can remove resolutions in package.json and run yarn && yarn dev, then visit http://localhost:4000.

### PR DESCRIPTION
**Pull request by theKashey**
fixes #832, breaks #775

The idea - in beta14 we introduced a new behavior for stand-in, to be more clear - returning an instance from stateless components.
This fixes #775, but actually required __only__ by 775/Relay, but in the same time breaks react-async-component and possible related stuff, which can not handle instances from return.

Thus we got a new configuration option - compat, default to true.
In compat mode we will NOT return an instance from stateless components, but just rendering them.
In case of instance - it will throw an error, asking to change configuration.

Instance as result is far rare case, and it possible to detect that you have it, that opposite. That's why it defaults to true.

Not listing a new option in readme, as long one should not configure it before hands.
** Future generations: if you got error "an instance were returned from a render function. Configuration change required" - you know the solution *

### CODE REVIEW

**Comment by taion** [comment]
cc @ctrlplusb 

Is it impossible to patch react-tree-walker to handle this case? It seems like it'd be a pretty odd limitation of react-tree-walker to not work with Relay at all.

**Comment by theKashey** [comment]
https://github.com/ctrlplusb/react-tree-walker/issues/19

**Comment by theKashey** [comment]
Anyway - I expect not only react-tree-walker to fail traverse "the tree". Even we are thinking to simplify traverse, dropping the odd cases and rely on sub-render(#830).

**Comment by taion** [comment]
Async is hard \:

Still, these trees seem traversable. Just another branch to handle for functional components.

File: docs/Troubleshooting.md [hunk 1]
```diff
@@ -311,3 +311,19 @@ import * as React from 'react' // React is now patched
...
+#### React-hot-loader: an instance were returned from a render function. Configuration change required
+
+React-hot-loader runs in "compatibility" mode by default. This disables some workarounds you possible might need.
+
+* **Enable** compat mode if you are using something like `react-async-bootstrapper`(react-async-component)
+* **Disable** compat mode if you are using `Relay`.
+
+There is **no** way to use them simultaneously.
+
+Buy default RHL will produce more simply and compatible code, but non-compact mode is absolutely legit for React 15+.
```
Comment (neoziro) [inline docs/Troubleshooting.md]: This is not very clear, I suggest to change everything and put it in the README :

```
#### Using React Hot Loader + Relay

To enable Relay compatibility mode, you must define a custom config:

__```js
import { setConfig } from 'react-hot-loader'
setConfig({ relayCompatMode: true })
__```

This mode is not compatible with `react-async-bootstrapper` or `react-async-component`, more detail in #833.
```

File: examples/styled-components/src/Counter.js [hunk 2]
```diff
@@ -15,7 +15,7 @@ class Counter extends React.Component {
   }
 
   render() {
-    return this.state.count
+    return <span>1:{this.state.count}</span>
```
Comment (neoziro) [inline examples/styled-components/src/Counter.js]: 1. ?

File: packages/react-hot-loader/src/patch.dev.js [hunk 3]
```diff
@@ -1,5 +1,6 @@
 import React from 'react'
 import reactHotLoader from './reactHotLoader'
+import './utils.dev'
```
Comment (neoziro) [inline packages/react-hot-loader/src/patch.dev.js]: `./utils.dev` should not do anything at runtime, refactoring needed.

File: packages/react-hot-loader/src/reactHotLoader.js [hunk 4]
```diff
@@ -74,6 +74,7 @@ const reactHotLoader = {
 
   config: {
     logLevel: 'error',
+    compat: true,
```
Comment (neoziro) [inline packages/react-hot-loader/src/reactHotLoader.js]: Rename to `relayCompatMode` (even if it not just `relay`, for now it is).

File: packages/react-stand-in/src/createClassProxy.js [hunk 5]
```diff
@@ -86,6 +87,21 @@ function createClassProxy(InitialComponent, proxyKey, wrapResult = identity) {
...
+        'React-hot-loader: ',
+        InitialComponent,
+        'returned an instance',
+        result,
+        'as result.',
+        'You have to disable compat mode to let React-hot-loader handle this.',
+        'setConfig({compat: false}))',
+      )
+      throw new Error(
+        'React-hot-loader: an instance were returned from a render function. Configuration change required',
```
Comment (neoziro) [inline packages/react-stand-in/src/createClassProxy.js]: This is React Stand In, not React Hot Loader.

File: packages/react-hot-loader/src/utils.dev.js [hunk 6]
```diff
@@ -1,9 +1,9 @@
+import { setConfig as setProxyConfig } from 'react-stand-in'
 import { getProxyByType } from './reconciler/proxies'
 import reactHotLoader from './reactHotLoader'
 import logger from './logger'
-import { setConfig as setProxyConfig } from 'react-stand-in'
 
-setProxyConfig({ logger })
+setProxyConfig({ logger, reactHotLoader: reactHotLoader.config })
```
Comment (neoziro) [inline packages/react-hot-loader/src/utils.dev.js]: If you are doing that, please remove `react-stand-in` and put all code in `react-hot-loader`. I think this is a good idea. We could also remove Lerna.

File: packages/react-stand-in/src/createClassProxy.js (hunk 5 above)
Comment (neoziro) [reply packages/react-stand-in/src/createClassProxy.js]: If it is, then we can remove `react-stand-in`.

File: docs/Troubleshooting.md (hunk 1 above)
Comment (theKashey) [reply docs/Troubleshooting.md]: We do know that "the hack" is needed only for Relay, but it is an absolutely legit structure from React prospective, and someone else could use it. Thus we should not call it `relay` even if we don't know that someone.

Comment (neoziro) [reply docs/Troubleshooting.md]: I agree, but could we find a name more precise than `compat`?

Comment (neoziro) [reply docs/Troubleshooting.md]: Maybe long and explicit, it is OK for me.

Comment (theKashey) [reply docs/Troubleshooting.md]: the "real" name is `supportIndeterminateComponent`, but actually not support, but "use".
`statelessIndeterminateComponent` ?

File: examples/styled-components/src/Counter.js (hunk 2 above)
Comment (theKashey) [reply examples/styled-components/src/Counter.js]: Eh, just some trash from smoke testing.

File: packages/react-hot-loader/src/utils.dev.js (hunk 6 above)
Comment (theKashey) [reply packages/react-hot-loader/src/utils.dev.js]: Let me refactor this a bit. 
Anyway - this is a good decision to make.

**Comment by theKashey** [comment]
This PR will cause merge conflict with sub-render one. It will be a good idea to cover proxy-related code with better tests after merging any of them.

**Comment by taion** [comment]
I'm still not sure this is a reasonable default, though. For example, if you look at npm, react-relay gets ~8x as many downloads a month as react-async-bootstrapper (88k v 11k).

As such, if this is working around a bug in react-async-bootstrapper anyway, then it probably shouldn't be the default behavior.

**Comment by theKashey** [comment]
Both of them actually are quite rarely used, but it is possible to detect relay and ask to change the configuration. We could detect "the instance was returned from a render". 
But we could not detect the opposite situation when that instance will blow up something outside our responsibility. And it impossible to understand that RHL is causing the issue.

**Comment by taion** [comment]
That's pretty reasonable.

I guess the only other factor is that Relay is an official Facebook thing. I dunno. Good point on being able to catch and log on errors w/Relay though.

**Comment by theKashey** [comment]
Maybe it worth invest a bit more time, and make RHL "function" compatible. Ie wrap functions with functions, not classes.

**Comment by theKashey** [comment]
I've spiked "class-less" solution, and here is outcome:
1. Everything works, except `hot-replacement-render`. Stateless components does not leave anything except `type` in traversable react tree. The common issue even for `enzyme` - there is no "instance"
2. I hoped that sub-render PR will mitigate this, and just "continue" hot-render on next class-based component.. but there might be no class-based component in next 10 steps.

Look like this is the only way to solve the puzzle for now.

**Comment by theKashey** [comment]
😿🤘☠️

**Comment by taion** [comment]
What happened?

**Comment by theKashey** [comment]
For now we ago going to support Relay, not React-async-component.

**Comment by taion** [comment]
This seemed like a pretty minimal diff, though. I was quibbling a bit regarding the default behavior, but this seemed like a pretty good and reasonable change.

**Comment by wKovacs64** [comment]
Thanks for all your work on RHL4, it's looking great.

I don't know about Relay or RAC, but I was waiting for this PR to land to fix Enzme shallow rendering of SFC's when using the babel plugin. Maybe I'm just doing something wrong, but this PR fixes it for me.

Simple repro based on RHL's webpack example: https://github.com/wKovacs64/rhl4-enzyme

Am I missing something obvious for this to work? Should I open a new issue? Thanks again.

**Comment by theKashey** [comment]
React-hot-loader should be disabled when you run stuff in jest.
Let me double check your repo.

**Comment by wKovacs64** [comment]
RHL should disable itself under Jest, or am I responsible for disabling it? In my real project, I currently only load the Babel plugin when env is development as a workaround for this issue, but I assumed that was a bit of a hack. Thanks for looking, let me know if I dorked something up!

**Comment by taion** [comment]
RHL doesn't do anything without `module.hot`, which can't possibly be defined in Jest, right?

I guess the alternative is to look for NODE_ENV as `test`?

Given the set of tools that this behavior has the potential to break, though, I guess I'm reluctantly moving toward the view that we Relay users should suffer the pain here 😔

**Comment by theKashey** [comment]
RHL will `completely` disable itself only in production, and `partially` without module.hot. Babel-plugin will be still active, but code it produces - shall not.

**Comment by wKovacs64** [comment]
I guess the Babel plugin being active without module.hot is what's behaving unexpectedly then (differently than RHL3, anyway). Should my demo repo work as-is, or should I disable Babel plugin when NODE_ENV=test?

**Comment by theKashey** [comment]
Our bad. During the last refactoring we just lose that rule about `module.hot`.

**Comment by wKovacs64** [comment]
No sweat, hope you didn't rush on my account. I just thought I just missed an answer in the discussion (I can be a little dense). 😄

**Comment by theKashey** [comment]
The fix was merged into next branch few minutes ago. Will be released soon.

**Comment by wKovacs64** [comment]
Sweet, thanks!

**Comment by theKashey** [comment]
Released among 4.0.0-beta.23

**Pull request closed without merging**